testcase/unittest/tools/test_config_gen.py

```diff
@@ -0,0 +1,236 @@
+import functools
+from typing import Callable
+
+import pytest
+
+from loopchain.tools.config_gen.config_gen import ConfigGenerator
+
+DEFAULT_KEY_PASSWORD = "pAsSwOrD"
+DEFAULT_TOTAL_REPS = 4
+DEFAULT_MAIN_REPS = 2
+DEFAULT_CHANNEL_NAMES = ("icon_dex", "dex_icon")
+
+
+@pytest.fixture
+def config_gen_factory(tmp_path) -> Callable[..., ConfigGenerator]:
+    def _config_gen_factory(_config_root_path, password=DEFAULT_KEY_PASSWORD,
+                            total_reps_count=DEFAULT_TOTAL_REPS, main_reps_count=DEFAULT_MAIN_REPS,
+                            channel_names=DEFAULT_CHANNEL_NAMES):
+        return ConfigGenerator(
+            config_root=_config_root_path,
+            key_password=password,
+            total_reps_count=total_reps_count,
+            main_reps_count=main_reps_count,
+            channel_names=channel_names
+        )
+
+    config_root_path = tmp_path / "config_root"
+    return functools.partial(_config_gen_factory, _config_root_path=config_root_path)
+
+
+class TestConfigGenerator:
+    def test_generate_all_stages(self, config_gen_factory):
+        config_gen = config_gen_factory()
+
+        config_gen.build()
+        config_gen.write()
+
+    def test_write_failed_when_config_already_exists(self, config_gen_factory):
+        config_gen = config_gen_factory()
+        config_gen.build()
+        config_gen.write()
+
+        with pytest.raises(FileExistsError):
+            config_gen.write()
+
+
+class TestConfigGeneratorKeys:
+    def test_build_all_reps_key(self, config_gen_factory):
+        expected_total_reps = 8
+        config_gen = config_gen_factory(total_reps_count=expected_total_reps)
+        assert not config_gen.keys
+
+        config_gen.build_keys()
+        assert config_gen.keys
+        assert expected_total_reps == len(config_gen.keys)
+
+    def test_build_returns_its_data_if_exists(self, config_gen_factory):
```

Comment:
What is the purpose of these methods below? Is it actually intended to set expected_data like test codes?

Reply:
I intended that `build_xxx` methods:
- build xxx if no data exists
- do nothing if xxx already exists as member.

Suppose that if someone wants to generate configure file with **exist key file**, not a generate new one. In this case, he or she is allowed to set his key file as `ConfigGenerator.keys`, then build rest of steps to get derived peer_configs etc.